PixelMaze/Player.js

```diff
@@ -0,0 +1,170 @@
+import { sec_to_frames, Tween } from "../MosaicSlider/Tween.js";
+import { Direction, to_y_down } from "../sketchlib/Direction.js";
+import { Index2D } from "../sketchlib/Grid.js";
+import { Tiles } from "./make_maze.js";
+import { Tilemap } from "./Tilemap.js";
+
+const TILE_SIZE = 16;
+
+const MovementState = {
+  IDLE: 0,
+  MOVING: 1,
+};
+Object.freeze(MovementState);
+
+// How fast to update the animation t value for sprites
+const ANIMATION_SPEED_SLOW = 0.1;
+const ANIMATION_SPEED_MEDIUM = 0.2;
+const ANIMATION_SPEED_FAST = 0.41;
+
+// How long the position tweens should take
+const DURATION_WALK = sec_to_frames(0.2);
+const DURATION_WALK_IN_PLACE = sec_to_frames(0.15);
+
+export class Player {
+  constructor(walk_sprites, idle_sprites, position) {
+    this.position = position;
+    this.direction = Direction.RIGHT;
+    this.walk_sprites = walk_sprites;
+    this.idle_sprites = idle_sprites;
+
+    this.state = MovementState.IDLE;
+
+    this.animation_t = 0;
+    this.animation_speed = ANIMATION_SPEED_SLOW;
+
+    this.input_direction = undefined;
+
+    this.movement_tween = undefined;
+
+    this.is_running = false;
+  }
+
+  handle_input(direction) {
+    this.input_direction = direction;
+  }
+
+  handle_run(run_button_pressed) {
+    this.is_running = run_button_pressed;
+  }
+
+  find_destination_cell(tilemap, position, direction) {
+    const { x, y } = position;
+    const cell_id = new Index2D(y / TILE_SIZE, x / TILE_SIZE);
+
+    const indices = tilemap.indices;
+    const neighbor = indices.get_neighbor(cell_id, direction);
+
+    // Check for obstacles in our path
+    if (neighbor === undefined || indices.get(neighbor) !== Tiles.FLOOR) {
+      return undefined;
+    }
+
+    // Move to the adjacent cell
+    const offset = to_y_down(direction).scale(TILE_SIZE);
+    return this.position.add(offset);
+  }
+
+  start_moving(tilemap, start_frame) {
+    const destination = this.find_destination_cell(
+      tilemap,
+      this.position,
+      this.input_direction
+    );
+
+    if (destination !== undefined) {
+      this.movement_tween = new Tween(
+        this.position,
+        destination,
+        start_frame,
+        DURATION_WALK
+      );
+      this.animation_speed = ANIMATION_SPEED_MEDIUM;
+    } else {
+      // Walk in place by setting start and end position to the current
```

Comment:
Ideally this should cue a wall-hit sound effect somehow. in #33 I'm experimenting with this using an `EventTarget`

Reply:
Mentioned this in notes